### PR TITLE
split recording session events and emitting audit events

### LINKED ISSUE #16773: Non SSH sessions are recorded even when session recording mode is off
Desktop, kube, DB, and app sessions are recorded even when the session recording mode is set to off, though the recordings are not easily playable in the web UI.

The root of this is that `events.AuditWriter` implements several different interfaces for session recording.

1. It implements `io.Writer`, by providing a `Write` method that converts the raw bytes to audit events which are then emitted to the underlying emitter. This is how SSH sessions are recorced.
2. It implements `events.Emitter` and exposes a `EmitAuditEvent` method directly. This is how desktop, DB, app, recordings work.

The problem is, the `RecordOutput` configuration option is only consulted in code path 1 (`io.Writer`). This means other types of sessions that call `EmitAuditEvent` directly do not check whether recording is enabled, and will always store the event, eventually writing it to disk.

Simply checking whenter `RecordOutput` is enabled before emitting the event is not a suitable solution, because this same interface is also used to emit session start/end and other audit events that we always want to see in the audit log.

This is yet another issue due to conflating emitting of audit events with session recording. We really should be treating them as separate features and not hiding them behind the same interface.

### LINKED ISSUE #25115: SSH events are written to to disk even when recording is disabled
Expected behavior:

No files are written to disk when session recording is disabled.

Current behavior:

"recordings" are written to <datadir>/log/records/<sid>.tar. The recordings contain 3 events (sessionstart, resize, and session.end).

Bug details:

- Teleport version: confirmed on v10.3.16, but likely present on all versions

Session recording is off (globally), and an external S3 recordings backend is not configured,.
```
auth_service:
  session_recording: off
```

This happens because `AuditWriter` only consults its `RecordOutput` configuration field for print events sent through the `io.Writer` interface. Any events submitted via the `EmitAuditEvent` interface will still be written to disk.

See #16773 for another related bug caused by this implementation.

### PR DESCRIPTION
**Pull request by capnspacehook**
**NOTE** many of the changes are renames or small fixes to existing tests.

This is a refactor of how audit events and session events are handled. Previously, all events were emitted using the same interface, `api/types/events.Emitter`. This lead to event-related code getting to be very confusing, as it was often unclear whether a given event was being recorded as a session event *and* emitted as an audit event, or only one of the two. Naturally, a few bugs arose due to this.

Using the comments in   as a guide, a separate interface for recording session events has been created. A `api/types/events.Recorder` should *only* be used to record session events, and an `Emitter` should now only be used to emit audit events. Instead of using a confusing `TeeWriter` that would transparently (and confusingly, given its name) hold a few event types that only belonged in session recordings, callers can now explicitly record and/or emit an event when necessary.

---

Manual testing I preformed:

For each item in the following list, I ensured events were recorded and/or emitted as necessary

- SSH:
  - Connect to node
  - Join/leave session
- Kube:
  - Connect to kube
- App Access:
  - Make requests to web app
- DB:
  - Login to database
  - Make queries to db
- Desktop Access:
  - Establish RDP session with a host

Merging this will break builds to `e`, but I'll have a small PR that will fix `e` builds ready before I merge this.

Fixes https://github.com/gravitational/teleport/issues/16773.
Fixes https://github.com/gravitational/teleport/issues/25115.

### CODE REVIEW

File: integration/integration_test.go [hunk 1]
```diff
@@ -4543,14 +4562,20 @@ func testAuditOff(t *testing.T, suite *integrationTestSuite) {
 	var err error
 
 	// create a teleport instance with auth, proxy, and node
+	sessionsDir := t.TempDir()
 	makeConfig := func() (*testing.T, []string, []*helpers.InstanceSecrets, *servicecfg.Config) {
+		auditConfig, err := types.NewClusterAuditConfig(types.ClusterAuditConfigSpecV2{
+			AuditSessionsURI: sessionsDir,
```
Comment (zmb3) [inline integration/integration_test.go:4568]: I'm surprised this works. Doesn't `AuditSessionsURI` have to be a URI? (`file:///home/log`)

File: lib/auth/helpers.go [hunk 2]
```diff
@@ -68,8 +69,10 @@ type TestAuthServerConfig struct {
 	// ClusterNetworkingConfig allows a test to change the default
 	// networking configuration.
 	ClusterNetworkingConfig types.ClusterNetworkingConfig
-	// Streamer allows a test to set its own audit events streamer.
+	// Streamer allows a test to set its own session events streamer.
```
Comment (zmb3) [inline lib/auth/helpers.go:72]:
```suggestion
	// Streamer allows a test to set its own session recording streamer.
```

File: lib/events/api.go [hunk 3]
```diff
@@ -799,10 +799,14 @@ type UploadMetadataGetter interface {
 	GetUploadMetadata(sid session.ID) UploadMetadata
 }
 
-// StreamWriter implements io.Writer to be plugged into the multi-writer
-// associated with every session. It forwards session stream to the audit log
-type StreamWriter interface {
+// SessionRecorder records session events. It can be used both as a
+// [io.Writer] when recording raw session data and as a [apievents.Recorder]
+// when recording session events.
+type SessionRecorder interface {
```
Comment (zmb3) [inline lib/events/api.go:805]: So we have both an `events.Recorder` interface and an `events.SessionRecorder` interface in this package, and they both are used for session recording? That's a bit confusing IMO.

File: lib/events/emitter.go [hunk 4]
```diff
@@ -614,13 +572,13 @@ func (s *CallbackStream) Complete(ctx context.Context) error {
 }
 
 // EmitAuditEvent emits audit event
```
Comment (zmb3) [inline lib/events/emitter.go:574]: Fix godoc here

File: lib/events/eventstest/mock_emitter.go [hunk 5]
```diff
@@ -30,6 +30,14 @@ type MockEmitter struct {
 	events []events.AuditEvent
 }
 
+func (e *MockEmitter) Write(_ []byte) (int, error) {
```
Comment (zmb3) [inline lib/events/eventstest/mock_emitter.go:33]:
```suggestion
func (e *MockEmitter) Write(b []byte) (int, error) {
```

Or `_` out `event` below. Just be consistent.

File: lib/events/recorder/recorder.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/gravitational/teleport/lib/events/filesessions"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// NewRecorder returns a [events.SessionRecorder]. If session recording is disabled,
+// a recorder is returned that will discard all session events. If session
+// recording is set to be synchronous, the returned recorder will use
+// syncStream to create an event stream. Otherwise, a streamer will be
+// used that will back recorded session events to disk for eventual upload.
+func NewRecorder(recCfg types.SessionRecordingConfig, cfg events.SessionWriterConfig, uploadDir string, syncStream events.Streamer) (events.SessionRecorder, error) {
```
Comment (zmb3) [inline lib/events/recorder/recorder.go:33]: Hmm, this feels like a broken abstraction. There's 3 possible outcomes here in the happy path:

1. we produce a discard stream because recording is disabled.
2. we are doing async recording to disk
3. we are streaming in sync mode

We have to pass an upload dir, even though it's only used in 1 of 3 scenarios, and we have to pass a sync streamer even though it's only used in 1 of 3 scenarios.

At this point, what good is the `NewRecorder` helper func if you have to create everything ahead of time and pass it in?

File: lib/events/session_writer.go [hunk 7]
```diff
@@ -76,15 +77,12 @@ type AuditWriterConfig struct {
...
-
 	// Component is a component used for logging
 	Component string
 
 	// MakeEvents converts bytes written via the io.Writer interface
 	// into AuditEvents that are written to the stream.
-	// For backwards compatibility, AuditWriter will convert bytes to
+	// For backwards compatibility, SessionWriter will convert bytes to
 	// SessionPrint events when MakeEvents is not provided.
 	MakeEvents func([]byte) []apievents.AuditEvent
```
Comment (zmb3) [inline lib/events/session_writer.go:87]: Some of the bugs we had with things not getting recorded correctly were due to us mixing the `io.Writer` adapter in with a general purpose session recorder.

I wonder if this would be simpler if `SessionWriter` didn't implement the `io.Writer` interface and only exposed the generic `RecordEvent` method. Then we could have a separate adapter for cases where we need `io.Writer`.

I'm thinking something like:

```
// ByteStreamRecorder records an arbitrary stream of bytes by converting
// them to SessionPrint events.
type ByteStreamRecorder struct {
    recorder events.Recorder
}

func (b *ByteStreamRecorder) Write(b []byte) (n int, err error) {
    // package b up into one or more SessionPrint events
    b.recorder.RecordEvent(printEvent)
}
```

File: lib/events/session_writer.go [hunk 8]
```diff
@@ -375,11 +370,33 @@ func (a *AuditWriter) EmitAuditEvent(ctx context.Context, event apievents.AuditE
 	}
 }
 
+func checkEventFields(event apievents.AuditEvent) error {
+	if err := checkBasicEventFields(event); err != nil {
+		return trace.Wrap(err)
+	}
+	if event.GetTime().IsZero() {
+		return trace.BadParameter("Time is unset, call SetupEvent before recording events")
```
Comment (zmb3) [inline lib/events/session_writer.go:378]: This is not a good message for a user - it doesn't tell them what's gone wrong or how to fix it.

File: lib/events/api.go [hunk 9]
```diff
@@ -799,10 +799,14 @@ type UploadMetadataGetter interface {
...
 	io.Writer
+	// SetupEvent will set necessary event fields, and must be called before
+	// the event is recorded.
+	SetupEvent(event apievents.AuditEvent) error
```
Comment (zmb3) [inline lib/events/api.go:809]: Does this have to be in the recorder interface?

With this, each recorder, whether it be a streaming recorder, one that writes to disk, a test one that just saves a copy in memory, or any future implementation is forced to implement the same logic for setting necessary fields.

Can we clean this up by either:
- moving the event setup upstream, so the recorder just records, and whatever calls the recorder is responsible for providing valid data
- wrapping with a checking streamer (or whatever its called now) in order to centralize this logic

File: lib/srv/app/server_test.go [hunk 10]
```diff
@@ -921,19 +915,19 @@ func TestRequestAuditEvents(t *testing.T) {
...
+		ServerEmitter: serverEmitter,
+		Apps:          types.Apps{app},
 	})
 
 	// make a request to generate events.
 	s.checkHTTPResponse(t, s.clientCertificate, func(_ *http.Response) {
 		// wait until chunk events are generated before closing the server.
 		require.Eventually(t, func() bool {
-			return chunkEventsReceived.Load() == 1
+			return chunkEventsReceived.Load() > 0
```
Comment (zmb3) [inline lib/srv/app/server_test.go:926]: What's the reasoning behind this change?

File: lib/srv/desktop/tdp/proto.go [hunk 11]
```diff
@@ -1469,8 +1467,10 @@ const tdpMaxNotificationMessageLength = 10240
 // The limit is kept as an additional defense-in-depth measure.
 const tdpMaxPathLength = 10240
 
-const maxClipboardDataLength = 1024 * 1024    // 1MB
-const tdpMaxFileReadWriteLength = 1024 * 1024 // 1MB
+const (
+	maxClipboardDataLength    = 1024 * 1024 // 1MB
+	tdpMaxFileReadWriteLength = 1024 * 1024 // 1MB
+)
```
Comment (zmb3) [inline lib/srv/desktop/tdp/proto.go:1473]: Guessing this was an auto-generated change?

We have several consts here, and two of them were combined. I'd revert the change or put all of the consts together.  Grouping some but not all of them makes it seems like these 2 values are in some way related, and I don't think that's actually the case.

File: lib/events/recorder/recorder.go (hunk 6 above)
Comment (zmb3) [inline lib/events/recorder/recorder.go:33]: The package is already called `recorder`, so how about `recorder.New` for a name?

File: lib/srv/sess.go [hunk 12]
```diff
@@ -922,6 +928,9 @@ func (s *session) emitSessionJoinEvent(ctx *ServerContext) {
 	}
 
 	// Emit session join event to Audit Log.
```
Comment (zmb3) [inline lib/srv/sess.go:930]: The comment now refers to the line below. I would move it down, or just remove it.

File: lib/srv/sess.go [hunk 13]
```diff
@@ -1240,43 +1255,42 @@ func (s *session) startTerminal(ctx context.Context, scx *ServerContext) error {
...
+		SessionID:   s.id,
+		Clock:       s.registry.clock,
+		Namespace:   s.serverMeta.ServerNamespace,
+		ServerID:    s.serverMeta.ServerID,
+		Component:   teleport.Component(teleport.ComponentSession, ctx.srv.Component()),
+		ClusterName: ctx.ClusterName,
+	}
+	// Return an SessionWriter with a discarding stream so closing it will
+	// work correctly for tests
+	if ctx.IsTestStub {
```
Comment (zmb3) [inline lib/srv/sess.go:1280]: I know you didn't write this, but I do wonder if any of your changes would make it easier to get rid of this. Having any sort of `if isTest` logic in production code is a smell.

**Review by zmb3 — COMMENTED** [review]
Definitely an improvement! I'm happy to see the confusing `TeeWriter` removed.

I'll give this a try tomorrow and see if it fixes some of the recording issues that I've run into.

I still find the sheer amount of types with similar names confusing. 

We have:

- Recorder
- Stream
- StreamManager
- SessionRecorder
- SessionWriter

These all kind of start to sound the same after a while.

File: lib/srv/desktop/tdp/proto.go (hunk 11 above)
Comment (capnspacehook) [reply lib/srv/desktop/tdp/proto.go:1473]: Yep that was a change from a formatter, I'll group all the consts in the same block

File: integration/integration_test.go (hunk 1 above)
Comment (capnspacehook) [reply integration/integration_test.go:4568]: I guess the scheme defaults to `file` if it's unset: https://github.com/gravitational/teleport/blob/capnspacehook/seperate-audit-record-events/api/utils/uri.go#L35

File: lib/events/recorder/recorder.go (hunk 6 above)
Comment (capnspacehook) [reply lib/events/recorder/recorder.go]: Good idea, done

File: lib/srv/sess.go (hunk 13 above)
Comment (capnspacehook) [reply lib/srv/sess.go]: Good point, I'll try and remove it

File: lib/srv/app/session.go [hunk 14]
```diff
@@ -286,51 +288,25 @@ func (s *Server) newStreamWriter(chunkID string) (events.StreamWriter, error) {
...
+		ClusterName: clusterName.GetClusterName(),
 	}
-
-	s.log.Debugf("Using async streamer for session chunk %v.", chunkID)
 	uploadDir := filepath.Join(
 		s.c.DataDir, teleport.LogsDir, teleport.ComponentUpload,
 		events.StreamingSessionsDir, apidefaults.Namespace,
 	)
-	fileStreamer, err := filesessions.NewStreamer(uploadDir)
+	streamWriter, err := recorder.New(recConfig, cfg, uploadDir, s.c.AuthClient)
```
Comment (greedy52) [inline lib/srv/app/session.go:304]: previous implementation using  `events.NewTeeStreamer` avoids sending `AppSessionRequestEvent` events to audit event (`AppSessionRequestEvent` only goes to session). I believe that logic now should be adjusted in `common.Audit`

Comment (capnspacehook) [reply lib/srv/app/session.go:304]: Thanks and good catch, fixed

File: lib/events/api.go (hunk 3 above)
Comment (capnspacehook) [reply lib/events/api.go:805]: Yeah that's a good point, I don't need both those interfaces now. Removed `Recorder` and `StreamManager`.

File: lib/events/session_writer.go (hunk 7 above)
Comment (capnspacehook) [reply lib/events/session_writer.go:87]: Hmm, that's not a bad idea, however I think this could be made more simple by returning different interfaces from the `recorder` package. One function could return a session recorder interface without `io.Writer`, and another function could wrap it and return an interface that includes `io.Writer`.

That would avoid creating another arbitrary type and keep all session event logic in SessionWriter, while also ensuring callers only get access to functionality that they need.
```suggestion
	MakeEvents func([]byte) []apievents.AuditEvent
```

File: lib/events/api.go (hunk 9 above)
Comment (capnspacehook) [reply lib/events/api.go:809]: The way it's implemented now, `SessionWriter` sets event indexes which requires keeping state. This code could be split from `SessionWriter`, but as we only have 2 recorders now that we have been exclusively for years (`SessionWriter` and `DiscardRecorder`), I think it's better to cross this bridge if we get to it.

File: lib/events/session_writer.go (hunk 8 above)
Comment (capnspacehook) [reply lib/events/session_writer.go:378]: I agree, though my goal for these errors is to let devs know when they forgot to call `SetupEvent` first. We really need some kind of debug-panic log or something, so we can panic and let developers know when they're not using something correctly without panicking when run in production.

File: lib/events/recorder/recorder.go (hunk 6 above)
Comment (capnspacehook) [reply lib/events/recorder/recorder.go]: I see what you mean, it's not a perfect solution by any means, but all services that record sessions duplicated this exact logic before. My thought was to centralize it somewhere to prevent new services from configuring a session recorder incorrectly, as even with all these changes handling events it not super straightforward.

File: lib/events/api.go (hunk 9 above)
Comment (capnspacehook) [reply lib/events/api.go:809]: The reason I split `SetupEvent` is because previously, `lib/events.SessionWriter` modified events in the caller's goroutine (called SetupEvent) and `lib/events.AsyncEmitter` read the same fields of the same events in a different goroutine, so events had to be recorded before they were emitted if doing both was desired to prevent races. These 2 structs are always used to record and emit respectively in production.

So centralizing this logic would be nice, but would open us up to easily adding races in the future by accident.

Comment (zmb3) [reply lib/events/api.go:809]: I still think we can fix the race without requiring a bunch of independent implementations of `SetupEvent`. Let me know if you want to brainstorm some solutions here,.

Comment (capnspacehook) [reply lib/events/api.go]: Updated to have a separate interface for setting up events.

File: lib/events/session_writer.go (hunk 7 above)
Comment (capnspacehook) [reply lib/events/session_writer.go:78]: After the changes to separate event setup and recording, doing this would necessitate adding 2 or more interfaces, a `SessionRecorder` without `io.Writer`, and `SessionSetterRecorder` without `io.Writer`. Not sure if this is worth making the event interface count even larger.

**Comment by capnspacehook** [comment]
@zmb3 @greedy52 ready for another look

File: lib/events/recorder/recorder.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+			return nil, trace.Wrap(err)
+		}
+		streamer = fileStreamer
+	}
+	swCfg.Streamer = streamer
+
+	rec, err := events.NewSessionWriter(swCfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Comment (greedy52) [inline lib/events/recorder/recorder.go:67]: curious if SessionWriter can/should implement events.SessionSetterRecorder as it owns a Setting from config anyway.

File: lib/srv/desktop/windows_server.go [hunk 16]
```diff
@@ -952,33 +919,40 @@ func (s *WindowsService) makeTDPSendHandler(ctx context.Context, emitter events.
 				// ones are around 2000 bytes. Anything approaching the limit of a single protobuf
 				// is likely some sort of DoS attempt and not legitimate RDP traffic, so we don't log it.
 				s.cfg.Log.Warnf("refusing to record %d byte PNG frame, image too large", len(b))
-			} else if err := emitter.EmitAuditEvent(ctx, e); err != nil {
-				s.cfg.Log.WithError(err).Warning("could not emit desktop recording event")
+			} else {
+				event, err := recorder.SetupSessionEvent(e)
+				if err == nil {
+					if err := recorder.RecordEvent(ctx, event); err != nil {
+						s.cfg.Log.WithError(err).Warning("could not record desktop recording event")
```
Comment (greedy52) [inline lib/srv/desktop/windows_server.go:926]: nit: not sure if worth it to have a `recorder.SetupAndRecordEvent` as this repeat a few places.

File: lib/events/session_writer.go [hunk 17]
```diff
@@ -281,17 +259,21 @@ func (a *AuditWriter) maybeSetBackoff(backoffUntil time.Time) bool {
...
+func (a *SessionWriter) RecordEvent(ctx context.Context, event apievents.AuditEvent) error {
+	if err := checkBasicEventFields(event); err != nil {
 		return trace.Wrap(err)
 	}
 
+	// the index starts at 0, so we'll only know if the index is invalid
+	// if we've seen at least one event already
+	if event.GetIndex() == 0 && a.acceptedEvents.Load() > 0 {
+		return trace.BadParameter("missing mandatory event index field")
+	}
```
Comment (greedy52) [inline lib/events/session_writer.go:272]: I assume no op setter will never be used with SessionWriter?

File: lib/srv/sess.go [hunk 18]
```diff
@@ -919,7 +925,9 @@ func (s *session) emitSessionJoinEvent(ctx *ServerContext) {
 		sessionJoinEvent.ConnectionMetadata.LocalAddr = ctx.ServerConn.LocalAddr().String()
 	}
 
-	// Emit session join event to Audit Log.
+	if err := s.recordEvent(ctx.srv.Context(), sessionJoinEvent); err != nil {
+		s.log.WithError(err).Warn("Failed to record session join event.")
+	}
```
Comment (greedy52) [inline lib/srv/sess.go:930]: i am assuming the reason `s.recordEvent` is before `s.emitAuditEvent` because recordEvent does the setup?

**Review by greedy52 — APPROVED** [review]

File: lib/events/session_writer.go (hunk 17 above)
Comment (capnspacehook) [reply lib/events/session_writer.go:272]: Yep a no-op setter should only be used for tests or when events will be discarded

File: lib/srv/sess.go (hunk 18 above)
Comment (capnspacehook) [reply lib/srv/sess.go:930]: Yes exactly

File: lib/events/setter.go [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	srv, ok := event.(ServerMetadataSetter)
+	if ok {
+		srv.SetServerNamespace(c.cfg.Namespace)
+		srv.SetServerID(c.cfg.ServerID)
+	}
+
+	event.SetIndex(int64(c.eventIndex.Load()))
+	c.eventIndex.Add(1)
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply lib/events/setter.go]: Good catch, fixed

File: lib/events/recorder/recorder.go (hunk 15 above)
Comment (capnspacehook) [reply lib/events/recorder/recorder.go:67]: Was on the fence about that, but I think it would help if it did. Pushed a commit that adds this

File: lib/srv/desktop/windows_server.go (hunk 16 above)
Comment (capnspacehook) [reply lib/srv/desktop/windows_server.go]: I think in places where the event will only be recorded this makes sense, done

File: lib/events/recorder/recorder.go [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/gravitational/teleport/lib/events/filesessions"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// New returns a [events.SessionSetterRecorder]. If session recording is disabled,
+// a recorder is returned that will discard all session events. If session
+// recording is set to be synchronous, the returned recorder will use
+// syncStream to create an event stream. Otherwise, a streamer will be
+// used that will back recorded session events to disk for eventual upload.
+func New(setterCfg events.SetterConfig, swCfg events.SessionWriterConfig, recCfg types.SessionRecordingConfig, uploadDir string, syncStream events.Streamer) (events.SessionSetterRecorder, error) {
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply lib/events/recorder/recorder.go:33]: `New` doesn't do too much that is true, but it does bundle the exact same (somewhat non-obvious) code used almost half a dozen times in the codebase into one place, making it easy to update in the future. 5 or so repetitions was enough to make me decide to put it into a function.

File: lib/events/api.go [hunk 21]
```diff
@@ -799,13 +799,28 @@ type UploadMetadataGetter interface {
...
+// [io.Writer] when recording raw session data and as a [apievents.Recorder]
+// when recording session events.
+type SessionRecorder interface {
 	io.Writer
 	apievents.Stream
 }
 
+// SessionSetterRecorder sets necessary session event fields and records them.
+type SessionSetterRecorder interface {
+	SetupAndRecordEvent(ctx context.Context, event apievents.AuditEvent) error
```
Comment (zmb3) [inline lib/events/api.go:819]: If a concrete type implements both `SessionEventSetter` and `SessionRecorder` doesn't it have everything it needs?

(Why do we need to require that it also implements `SetupAndRecordEvent`?)

File: lib/events/api.go [hunk 22]
```diff
@@ -799,13 +799,28 @@ type UploadMetadataGetter interface {
...
+	SessionEventSetter
+	SessionRecorder
+}
+
 // StreamEmitter supports submitting single events and streaming
 // session events
 type StreamEmitter interface {
```
Comment (zmb3) [inline lib/events/api.go:826]:
```suggestion
// StreamEmitter supports emitting single events to the audit log
// and streaming events to a session recording.
type StreamEmitter interface {
```

File: lib/events/discard.go [hunk 23]
```diff
@@ -64,38 +66,60 @@ func (d *DiscardAuditLog) StreamSessionEvents(ctx context.Context, sessionID ses
...
 }
 
-// Complete does nothing
-func (*DiscardStream) Complete(ctx context.Context) error {
+// Complete marks the stream as closed
+func (d *DiscardRecorder) Complete(ctx context.Context) error {
+	if !d.completed.Load() {
+		close(d.done)
+	}
+	d.completed.Store(true)
```
Comment (zmb3) [inline lib/events/discard.go:113]: You'll want to use a compare and swap here, otherwise it's possible that the close call happens twice.
```suggestion
	if !d.completed.CompareAndSwap(false, true) {
		close(d.done)
	}
```

File: lib/events/setter.go [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+type SetterAndRecorder struct {
+	SessionEventSetter
+	SessionRecorder
+}
+
+// SetupAndRecordEvent will set necessary event fields for session-related
+// events and record them.
+func (s SetterAndRecorder) SetupAndRecordEvent(ctx context.Context, event apievents.AuditEvent) error {
```
Comment (zmb3) [inline lib/events/setter.go:138]: So if you removed `SetupAndRecordEvent` from `SessionSetterRecorder`, then this could just be a free function:

```
func SetupAndRecord(ctx context.Context, s SessionSetterRecorder, event apievents.AuditEvent) error {
    // the body remains the same, but now you no longer
    // need a new struct for SetterAndRecorder
}
```

File: lib/kube/proxy/sess.go [hunk 25]
```diff
@@ -561,10 +564,16 @@ func (s *session) launch() error {
...
+	})
+	if err == nil {
+		if err := s.recorder.RecordEvent(s.forwarder.ctx, sessionStartEvent); err != nil {
+			s.forwarder.log.WithError(err).Warn("Failed to record session start event.")
+		}
+		if err := s.emitter.EmitAuditEvent(s.forwarder.ctx, sessionStartEvent); err != nil {
+			s.forwarder.log.WithError(err).Warn("Failed to emit session start event.")
+		}
+	} else {
+		s.forwarder.log.WithError(err).Warn("Failed to setup session start event.")
```
Comment (zmb3) [inline lib/kube/proxy/sess.go:576]:
```suggestion
		s.forwarder.log.WithError(err).Warn("Failed to set up session start event - event will not be recorded")
```

File: lib/srv/sess.go [hunk 26]
```diff
@@ -1238,43 +1252,44 @@ func (s *session) startTerminal(ctx context.Context, scx *ServerContext) error {
...
+			Context:   ctx.srv.Context(),
+			SessionID: s.id,
+			Clock:     s.registry.clock,
+			Component: teleport.Component(teleport.ComponentSession, ctx.srv.Component()),
+		},
+		ctx.SessionRecordingConfig,
+		filepath.Join(
+			ctx.srv.GetDataDir(), teleport.LogsDir, teleport.ComponentUpload,
+			events.StreamingSessionsDir, ctx.srv.GetNamespace(),
+		),
```
Comment (zmb3) [inline lib/srv/sess.go:1283]: This logic for building the directory path for a recording seems to be duplicated in a few places - what do you think about a utility in `lib/events` to consolidate this?

**Review by zmb3 — COMMENTED** [review]
We're nearly there - a few minor comments while I give this a test drive.

**Comment by zmb3** [comment]
I tried to launch a desktop session and was presented with this error:

image

File: lib/events/discard.go (hunk 23 above)
Comment (capnspacehook) [reply lib/events/discard.go]: Thanks, done

File: lib/events/setter.go (hunk 24 above)
Comment (capnspacehook) [reply lib/events/setter.go]: That's a good idea, done

File: lib/srv/sess.go (hunk 26 above)
Comment (capnspacehook) [reply lib/srv/sess.go]: I rolled that into `lib/events/recorder.New` instead and simplified how `recorder.New` is called

File: lib/events/api.go (hunk 21 above)
Comment (capnspacehook) [reply lib/events/api.go]: This was a bit of a smell, making `SetupAndRecordEvent` a function is a much better idea

**Comment by capnspacehook** [comment]
Did another round of manual testing and ensured that error is fixed @zmb3.

File: lib/events/api.go [hunk 27]
```diff
@@ -799,15 +799,29 @@ type UploadMetadataGetter interface {
...
 
-// StreamWriter implements io.Writer to be plugged into the multi-writer
-// associated with every session. It forwards session stream to the audit log
-type StreamWriter interface {
+// SessionEventSetter will set necessary event fields for session-related
+// events and must be called before the event is recorded, regardless
+// of whether the event will be recorded, emitted, or both.
+type SessionEventSetter interface {
+	SetupSessionEvent(event apievents.AuditEvent) (apievents.AuditEvent, error)
+}
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply lib/events/api.go]: Addressed both suggestions. I had thought before about somehow enforcing that event setup is done before recording but this seems like a clear way to do it. I opted to make `PreparedSessionEvent` a new interface so it can't accidentally be misused by simply casting a `AuditEvent`. `SetupSessionEvent` being separate is a necessary evil unfortunately.

Previously `SessionWriter` would set event fields in the caller's goroutine, and `AsyncEmitter` would check the same event fields in a new goroutine when events were recorded or emitted respectively. It was very easy to trigger a data race by emitting and recording an event in that order. Making event setup separate means the same event can be recorded and emitted in either order with no problems.

File: integration/integration_test.go [hunk 28]
```diff
@@ -4736,6 +4763,52 @@ func testAuditOff(t *testing.T, suite *integrationTestSuite) {
...
+			hasLeave = true
+		case *apievents.SessionEnd:
+			hasEnd = true
+		default:
+			continue
+		}
+	}
+	require.True(t, hasStart)
+	require.True(t, hasLeave)
+	require.True(t, hasEnd)
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply integration/integration_test.go]: Done

**Comment by capnspacehook** [comment]
@nklaassen @zmb3 ready for another look

File: lib/events/setter.go [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+		cfg.Clock = clockwork.NewRealClock()
+	}
+	if cfg.UID == nil {
+		cfg.UID = utils.NewRealUID()
+	}
+
+	return nil
+}
+
+type Setter struct {
```
Comment (zmb3) [inline lib/events/setter.go:84]: missing godocs

**Review by zmb3 — APPROVED** [review]
I can confirm this fixes the desktop recording/playback issue.

File: lib/events/api.go [hunk 30]
```diff
@@ -799,15 +799,29 @@ type UploadMetadataGetter interface {
...
 }
 
-// StreamWriter implements io.Writer to be plugged into the multi-writer
-// associated with every session. It forwards session stream to the audit log
-type StreamWriter interface {
+// SessionEventSetter will set necessary event fields for session-related
+// events and must be called before the event is recorded, regardless
+// of whether the event will be recorded, emitted, or both.
+type SessionEventSetter interface {
+	SetupSessionEvent(event apievents.AuditEvent) (apievents.PreparedSessionEvent, error)
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply lib/events/api.go]: Good point, done

File: api/types/events/api.go [hunk 31]
```diff
@@ -77,11 +77,19 @@ type Emitter interface {
 	EmitAuditEvent(context.Context, AuditEvent) error
 }
 
+// PreparedSessionEvent is an event that has been prepared by
+// a [github.com/gravitational/teleport/lib/events.SessionEventSetter].
+// More specifically, it is a wrapper around an AuditEvent that signifies
+// the event has been prepared and is ready to be recorded or emitted.
+type PreparedSessionEvent interface {
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply api/types/events/api.go:84]: I wanted to avoid a exported concrete type so devs couldn't inadvertently create a prepared event without actually preparing it. The way it is now you have to use a preparer from `lib/events` or refactor the existing code so you can create a concrete prepared event yourself which would be a red flag.

[the remark this answers is not in the document]
Comment (zmb3) [reply api/types/events/api.go:84]: Yeah, I agree with that. 

Maybe this is one good use case for an exported method that returns an unexported type.

[the remark this answers is not in the document]
Comment (nklaassen) [reply api/types/events/api.go:84]: I may be misremembering but I think our linter will whine about an exported method that returns an unexported type.

I think exporting a concrete type is fine, if a caller is willing to make the explicit cast, they are saying they know what they are doing, they can deal with the consequence. The interface kind of gets in the way, there's at least one place already where you create a nooppreparer just to convert the type, that's a perfect place to just do the type cast I think

**Review by nklaassen — APPROVED** [review]
I like the `PreparedSessionEvent` changes, makes it clearer to me what you need to call when. I think it may be better as a concrete type, but the interface is alright

File: lib/events/eventstest/mock_emitter.go [hunk 32]
```diff
@@ -1,90 +0,0 @@
...
-import (
-	"context"
-	"sync"
-
-	"github.com/gravitational/teleport/api/types/events"
-	"github.com/gravitational/teleport/lib/session"
-)
-
-// MockEmitter is an emitter that stores all emitted events.
-type MockEmitter struct {
```
Comment (codingllama) [inline lib/events/eventstest/mock_emitter.go:28]: Should we keep this type as an alias, or just keep the impl itself, while we do the e/ changes? Won't this break the build otherwise?

Comment (codingllama) [reply lib/events/eventstest/mock_emitter.go:28]: Same for other breaking changes.

Comment (capnspacehook) [reply lib/events/eventstest/mock_emitter.go:28]: Good point, I'll make an alias

File: api/types/events/api.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (zmb3) [reply api/types/events/api.go:84]: Just spitballing here, not even sure I like this idea...

You would have to move some things around - the code that creates prepared events would have to live in api/events instead of lib/events, but then you could have something like:

```
type preparedEvent struct { ... }

func PrepareEvent(event AuditEvent) preparedEvent { ... }

func Record(event preparedEvent) { ... }
```

Normally I'd push back against such a change, as it's kind of a pain to work with public APIs that require unexported types. In this case, it's kind of what you want though: there's only one way to get a `preparedEvent`, so you're forced into doing the right thing and calling `PrepareEvent` first.

If we export `PreparedEvent`, a new dev struggling to get their code to compile might just add an explicit type conversion and move on, not realizing that they broke the rules. (Though maybe Nic's right and this isn't a big deal)

This would also have impacts on your tests, because if you wanted to keep a no-op preparer, it would have to be in the api/events package.

What do you guys think? I'm still not sure, which probably means we shouldn't do it.

[the remark this answers is not in the document]
Comment (capnspacehook) [reply api/types/events/api.go:84]: Ah I understand what you mean now. I had the exact same idea, but you're unfortunately right in that it won't work inside `api/events`. Too much code would need `preparedEvent` inside `lib/events` which it wouldn't have access to. I tried moving `Stream` to `lib/events` so `preparedEvent` could live there as well, but that would require `api/events` importing `lib/events` which is a non-starter.

**Comment by codingllama** [comment]
Thanks for putting in the effort to not break e/. Much appreciated!
